File: coco/src/test/java/coco/ep/M7Test.java
package coco.ep;

import org.junit.Assert;
import org.junit.Test;

public class M7Test {

    public interface TestTemplate<FT> extends coco.ep.m7.Factory<FT>, M6Test.TestTemplate<FT> {
        default void test() {
            M6Test.TestTemplate.super.test();

            Exp<FT> pby = convert(this.lit(3.0)).powby(this.lit(2.0));
            Exp<FT> d1 = convert(this.divd(this.lit(5.0), this.lit(2.0))).powby(this.lit(2.0));
            Exp<FT> m1 = convert(this.mult(this.lit(5.0), this.lit(2.0))).powby(this.lit(2.0));
            Exp<FT> a1 = convert(this.add(this.lit(5.0), this.lit(2.0))).powby(this.lit(2.0));
            Exp<FT> s1 = convert(this.sub(this.lit(5.0), this.lit(3.0))).powby(this.lit(2.0));
            Exp<FT> n1 = convert(this.neg(this.lit(5.0))).powby(this.lit(2.0));

            Assert.assertEquals(9.0, convert(pby).eval(), 0.0);
            Assert.assertEquals(6.25, convert(d1).eval(), 0.0);
            Assert.assertEquals(100.0, convert(m1).eval(), 0.0);
            Assert.assertEquals(49.0, convert(a1).eval(), 0.0);
            Assert.assertEquals(4.0, convert(s1).eval(), 0.0);
            Assert.assertEquals(25.0, convert(n1).eval(), 0.0);

            Assert.assertEquals(28561.0, convert(convert(this.add(this.lit(1.0), this.lit(12.0))).powby(this.lit(4.0))).eval(), 0.0);
        }
    }
    private static class ActualTest implements M7Test.TestTemplate<coco.ep.m7.finalized.Exp>, coco.ep.m7.finalized.Factory {}

    @Test
    public void testTest() { new M7Test.ActualTest().test(); }
}

File: coco_scala/src/test/scala/exp/m7/M7Test.scala
package exp.m7

import org.scalatest.FunSuite

trait TestTemplate[FT] extends Factory[FT] with exp.m4.TestTemplate[FT] {
  val suite:FunSuite
  import suite._

  override def test() : Unit = {
    super.test()

    val pby = this.lit(3.0).powby(this.lit(2.0));
    val d1 = this.divd(this.lit(5.0), this.lit(2.0)).powby(this.lit(2.0));
    val m1 = this.mult(this.lit(5.0), this.lit(2.0)).powby(this.lit(2.0));
    val a1 = this.add(this.lit(5.0), this.lit(2.0)).powby(this.lit(2.0));
    val s1 = this.sub(this.lit(5.0), this.lit(3.0)).powby(this.lit(2.0));
    val n1 = this.neg(this.lit(5.0)).powby(this.lit(2.0));

    assert(9.0 === pby.eval)
    assert(6.25 === d1.eval)
    assert(100.0 === m1.eval)
    assert(49.0 === a1.eval)
    assert(4.0 === s1.eval)
    assert(25.0 === n1.eval)

    assert(28561.0 === this.add(this.lit(1.0), this.lit(12.0)).powby(this.lit(4.0)).eval)
  }
}

class M7Test extends FunSuite { self =>

  object ActualTest extends TestTemplate[exp.m7.finalized.Exp] with finalized.Factory {
    val suite: FunSuite = self
  }

  test("M7") {
    ActualTest.test()
  }
}
